mmdet/datasets/base_det_dataset.py

```diff
@@ -0,0 +1,107 @@
+from mmengine.fileio import FileClient, load
+from mmengine.utils import is_abs
+
+from ..registry import DATASETS
+
+
+@DATASETS.register_module()
+class BaseDetDataset(BaseDataset):
+    """Base dataset for detection.
+
+    Args:
+        proposal_file (str, optional): Proposals file path. Defaults to None.
+        file_client_args (dict): Arguments to instantiate a FileClient.
+            See :class:`mmengine.fileio.FileClient` for details.
+            Defaults to ``dict(backend='disk')``.
+    """
+
+    def __init__(self,
+                 *args,
+                 proposal_file: Optional[str] = None,
+                 file_client_args: dict = dict(backend='disk'),
+                 **kwargs) -> None:
+        self.proposal_file = proposal_file
+        self.file_client_args = file_client_args
+        self.file_client = FileClient(**file_client_args)
+        super().__init__(*args, **kwargs)
+
+    def full_init(self) -> None:
+        """Load annotation file and set ``BaseDataset._fully_initialized`` to
+        True.
+
+        If ``lazy_init=False``, ``full_init`` will be called during the
+        instantiation and ``self._fully_initialized`` will be set to True. If
+        ``obj._fully_initialized=False``, the class method decorated by
+        ``force_full_init`` will call ``full_init`` automatically.
+
+        Several steps to initialize annotation:
+
+            - load_data_list: Load annotations from annotation file.
+            - load_proposals: Load proposals from proposal file, if
+              `self.proposal_file` is not None.
+            - filter data information: Filter annotations according to
+              filter_cfg.
+            - slice_data: Slice dataset according to ``self._indices``
+            - serialize_data: Serialize ``self.data_list`` if
+            ``self.serialize_data`` is True.
+        """
+        if self._fully_initialized:
+            return
+        # load data information
+        self.data_list = self.load_data_list()
+        # get proposals from file
+        if self.proposal_file is not None:
+            self.load_proposals()
+        # filter illegal data, such as data that has no annotations.
+        self.data_list = self.filter_data()
+
+        # Get subset data according to indices.
+        if self._indices is not None:
+            self.data_list = self._get_unserialized_subset(self._indices)
+
+        # serialize data_list
+        if self.serialize_data:
+            self.data_bytes, self.data_address = self._serialize_data()
+
+        self._fully_initialized = True
+
+    def load_proposals(self) -> None:
+        """Load proposals from proposals file.
+
+        The `proposals_list` should be a dict[img_path: proposals]
+        with the same length as `data_list`. And the `proposals` should be
+        a `dict` or :obj:`InstanceData` usually contains following keys.
+
+            - bboxes (np.ndarry): Has a shape (num_instances, 4),
```

Comment:
Why the bbox in the instance is a list of float while the proposal is a ndarray? Should they be consistent?

Reply:
The boxes will be processed in LoadAnnotation and returned as np.ndarry. I think this is mainly because each dataset have different annotation type, it is hard to uniform. 
But proposals can be processed durning DumpProposalMetric, which we do not need to convert to list and re-convert back to np.ndarry.

Reply:
We can discuss this in the PR of Fast R-CNN

Reply:
the instance is a list but proposals is an `InstanceData`, they are not same. Also, the proposal will not be processed in dataset, maybe we can uniform the proposal type.